enable use of two txqs in TT-Fabric on Blackhole

SeanNijjar opened the pull request:
With this change, both TXQ0 and TXQ1 are used by Blackhole. Sender channel traffic (payloads and departing credits) are mapped to TXQ0 and receiver channel traffic (departing ack/completion credits) are mapped to TXQ1.

The implementation is sub-optimal but is due to a HW bug in Blackhole Ethernet subsystem. The bug prevents concurrent Ethernet remote register writes across multiple TXQs. Specifically, on the receive side, register writes over Ethernet can only be (statically) directed to one RXQ at a time, configured through a register write. The register writes cannot be dynamically be directed to different receivers, based on sender or payload contents.

As a work-around, specifically for Blackhole in two TXQ operating mode, the receiver -> send credit scheme moves to counter based credits. These credits are sent using using the Ethernet send packet path (i.e. not a register write) . This approach is race-safe. The reason this approach is race-safe is because we use unbounded counters, as opposed to credits that indicate an increment of some kind. Since it's an unbounded counter, where the actual value is used as is directly, then it is okay if a value is sent multiple times. A value could be sent multiple times if the Erisc updates the values multiple times before the ethernet subsystem is able to dispatch the write over Ethernet. 

Note that in the abstract, there is a roll-over/overflow concern for unbounded counters. In practice, this is not an issue because of how subtraction is implemented in hardware. If the sender channel is comparing two unsigned counters: one that has wrapped around against one that hasn't. From the uint counter perspective, the wrap doesn't matter. A snippet such as the following running on the risc demonstrates it.

```
    uint32_t trailing = std::numeric_limits<uint32_t>::max() - 3;
    uint32_t leading = 4; // indicating that leading wrapped recently (current offset is 4 past wrap)

    uint32_t delta = leading - trailing;
    
    uint32_t counted_delta = 0;
    for (uint32_t ctr = trailing; ctr != leading; ctr++)  {
        counted_delta++;
    }

    assert(counted_delta == delta);
```

In addition to integration into the TT-Fabric router, this approach is implemented in the unit test for basic bringup and validation.

Additionally, a unit test to program this two-txq mode is added to test the programming sequence to set TXQ1 to packet mode in isolation. This unit test also loosely models the general traffic patterns sent by fabric routers.

To support the new behaviour, the fabric router code was updated to start formalizing these credit receivers/senders into separate types which are compile time selectable now.

```
                          ┌─── now a compile time configurable type                 
                          │                          ───┐                           
               Chip 0     │                             │      Chip 1               
            Fabric Router │            srcId            │   Fabric Router           
   ┌──────────────────────┼───────┐     │          ┌────┼─────────────────────────┐ 
   │                      │       │     │          │    │                         │ 
   │                      ▼       │    \│          │    ▼                         │ 
   │ ┌──────────────┐──────────┐  │   │ \          │ ┌────────┬────────────────┐  │ 
   │ │ Inbound 0    │ Credit   │  │   │  \ Credit  │ │Credit  │ Inbound 0      │  │ 
   │ │ Channel      │ Receiver │◄─┼───┤  |     ┌───┼─┤Sender  │ Channel        │  │ 
   │ └───────┬──────┘──────────┘  │   │  |◄────┘   │ └────────┴────────────────┘  │ 
 ┌─┼─────────┘                    │ ┌─┤  |         │                  ▲           │ 
 │ │ ┌──────────────┐──────────┐  │ │ │  /         │                  │           │ 
 │ │ │ Inbound 1    │ Credit   │  │ │ │ /          │                  │           │ 
 │ │ │ Channel      │ Receiver │◄─┼─┘  /           │                  │           │ 
 │ │ └───────┬──────┘──────────┘  │◄──────────────►│                  │           │ 
 │ │         │                    │     Ethernet   │                  │           │ 
 │ │         │                    │                │                  │           │ 
 │ │         │                    │                │                  │           │ 
 └─┼────────►└────────────────────┼────────────────┼──────────────────┘           │ 
   │                              │                │                              │ 
   │                              │                │                              │ 
   │                              │                │                              │ 
   │                              │                │                              │ 
   │                              │                │                              │ 
   └──────────────────────────────┘                └──────────────────────────────┘ 
```

In this change, the direction of credit send/receive (relative to Ethernet link) still hardcoded by type but in the future they will likely be generalized further to support various interfaces (Ethernet-subsystem, noc). 



### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI: https://github.com/tenstorrent/tt-metal/actions/runs/16209520690
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable): https://github.com/tenstorrent/tt-metal/actions/runs/16243051049
- [x] BH LLMbox: https://github.com/tenstorrent/tt-metal/actions/runs/16243041071
  - note on reruns: Each run passed. Wanted to confirm basic stability/robustness. Also ran locally fabric tests locally 200x without failures
- [x] BH Nightly: https://github.com/tenstorrent/tt-metal/actions/runs/16243050111
- [x] T3K Unit Tests: https://github.com/tenstorrent/tt-metal/actions/runs/16206003333
- [x] Metal microbenchmarks: https://github.com/tenstorrent/tt-metal/actions/runs/16209524287
- [X] New/Existing tests provide coverage for changes

SeanNijjar commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -76,20 +76,47 @@ FabricEriscDatamoverConfig::FabricEriscDatamoverConfig(Topology topology) {
     this->edm_status_address = edm_local_sync_address + field_size;
 
     uint32_t buffer_address = edm_status_address + field_size;
-    this->num_riscv_cores = tt::tt_metal::MetalContext::instance().hal().get_processor_classes_count(
```
moved up before we start traversing the L1 addresses (this doesn't do address traversal)

SeanNijjar commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_txq_setup.h:
```diff
@@ -0,0 +1,133 @@
+constexpr uint64_t TXQ0_MAC_ADDR = BCAST_MAC_ADDR;
+constexpr uint64_t TXQ1_MAC_ADDR = MCAST_MAC_ADDR;
+
+#define ETH_CORE_A_ETH_CTRL_A_MAC_RX_ADDR_ROUTING_REG_ADDR (0xFFB98154)
```
Note I pulled in the minimal macros/values in as possible because the files these are sourced from contain many more registers and memory layout information and it's not clear yet if they _all_ can be mainlined. If it turns out they can be, we'll push the whole register map file and include it here and remove the local defines, at a later date.

SeanNijjar commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_txq_setup.h:
```diff
@@ -0,0 +1,133 @@
+
+static void eth_enable_packet_mode(uint32_t qnum) {
+    // Initializing the rx queue to enable packet mode
+    //   eth_rxq_init(qnum, (TMP_8K_BUF_ADDR + MAX_ETH_PACKET_SIZE) >> 4, MAX_ETH_PACKET_SIZE >> 4, false,  true);
```
going to delete

abhullar-tt commented on tests/tt_metal/tt_metal/common/multi_device_fixture.hpp:
```diff
@@ -58,6 +58,26 @@ class N300DeviceFixture : public DispatchFixture {
     }
 };
 
+class P300DeviceFixture : public DispatchFixture {
```
this will also run on P150, could you rename

abhullar-tt commented on tests/tt_metal/tt_metal/eth/test_eth_multi_txq_rxq.cpp:
```diff
@@ -0,0 +1,176 @@
+    for (const auto& eth_core : active_eth_cores) {
+        chip_id_t connected_device_id;
+        CoreCoord connected_eth_core;
+        std::tie(connected_device_id, connected_eth_core) = device_0->get_connected_ethernet_core(eth_core);
```
does this API account for an active eth core where link could be down? there is a cluster api that checks if link is down (see this a lot on non-reworked boards)

SeanNijjar replied:
I see - I'll double check if that's the case or not

abhullar-tt commented on tt_metal/fabric/hw/inc/edm_fabric/fabric_txq_setup.h:
```diff
@@ -0,0 +1,130 @@
+    // For TXQ0, we are we are setting the destination mac address to 0 (RXQ0)
+    //  -> these will map to "unicast" packets
+    // For TXQ1, we are we are setting the destination mac address to 1 (RXQ1)
+    //  -> these will map to "bcast"/"mcast" packets
```
the 2 risc protocol for Fabric has each risc handling different type of packets (unicast/mcast)? or one risc handles sender and other handles receiver and they have statically assigned queues?

SeanNijjar replied:
yes to both:
- the type of packet is used by the mac/eth subsystem to decide which rxq the packet gets routed to
- the sender and receiver directions in fabric are statically mapped (CT args) to a different TXQ

SeanNijjar replied:
you assign it on both ends, so for example if TXQ0 is programmed to send MCAST, then RXQ0 would be programmed to have MCAST routed to it. Then for example, we might choose to do the same for BCAST through TXQ1 and RXQ1

abhullar-tt commented on tt_metal/fabric/hw/inc/edm_fabric/fabric_txq_setup.h:
```diff
@@ -0,0 +1,130 @@
+    enable_packet_mode_rxq(ETH_CORE_A_ETH_RXQ_0_CTRL_REG_ADDR);
+    enable_packet_mode_rxq(ETH_CORE_A_ETH_RXQ_1_CTRL_REG_ADDR);
+
+    map_tx_queues_to_rx_queues();
```
do the registers being set here need to be cleared on teardown?

SeanNijjar replied:
also in general, I was told these registers can generally be programmed "any time" so long as there aren't active traffic flows through those reprogrammed queues, during the time of programming

abhullar-tt replied:
if its managed by boot FW it would limit the reprogrammability of the queues? but I guess we keep it in one mode for the life of the device usage

SeanNijjar replied:
my understanding is boot FW switches TXQ0 to packet mode (I don't really change that). The values I set for it were based off of what I saw it being programmed to coming out of reset - I kept it "unchanged" in that sense.

TXQ1 is not really touched by it, and TXQ2 is exclusively raw mode (I believe it doesn't even support packet mode in hardware)

aagarwalTT commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -76,20 +76,47 @@ FabricEriscDatamoverConfig::FabricEriscDatamoverConfig(Topology topology) {
+        this->risc_configs.emplace_back(risc_id);
+    }
+
+    log_info(tt::LogFabric, "sender_txq_id: {}, receiver_txq_id: {}", this->sender_txq_id, this->receiver_txq_id);
```
I suppose this is only here for local debugging?

SeanNijjar replied:
will delete

aagarwalTT commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -396,6 +423,9 @@ FabricEriscDatamoverConfig::FabricEriscDatamoverConfig(
     FabricEriscDatamoverConfig(topology) {
     this->sender_txq_id = 0;
     this->receiver_txq_id = 0;
+    if (tt::tt_metal::MetalContext::instance().hal().get_arch() == tt::ARCH::BLACKHOLE) {
+        this->receiver_txq_id = 1;
+    }
```
seems duplicated? we are already setting this in `FabricEriscDatamoverConfig(topology)`

aagarwalTT commented on tt_metal/fabric/hw/inc/edm_fabric/1d_fabric_constants.hpp:
```diff
@@ -344,7 +346,62 @@ static_assert(
+        NUM_RECEIVER_CHANNELS>();
+
+template <typename T>
+constexpr bool counter_credit_addresses_are_valid(const T& counter_addresses) {
```
just curious- in what cases could these addresses be invalid?

SeanNijjar replied:
these are not set if only one Eth TXQ is in use (or atleast they are allowed to be unset)

aagarwalTT commented on tt_metal/fabric/hw/inc/edm_fabric/fabric_router_flow_control.hpp:
```diff
@@ -0,0 +1,214 @@
+#include "tt_metal/hw/inc/ethernet/tt_eth_api.h"
+#include "tt_metal/hw/inc/ethernet/tunneling.h"
+
+struct ReceiverChannelCounterBasedResponseCreditSender {
```
Just wondering if we could drop `ReceiverChannel`/`SenderChannel` from these names just to make them more manageable and leave it implied that receiver channel will be the credit sender and sender channel will be the credit receiver?

SeanNijjar replied:
My preference was to make it very explicit for now because the implementations of each only support one direction. I would like these to eventually be more generalized to support different directions and interfaces (noc vs eth) and I think at that point we should totally change up the naming.

If you have any better names than the one chosen in the mean time, I'm definitely open to changing them.

daminakaTT commented on tests/tt_metal/tt_metal/eth/test_eth_multi_txq_rxq.cpp:
```diff
@@ -0,0 +1,176 @@
+        t2.join();
+    }
+
+    return pass;
```
looks like always true?

daminakaTT commented on tests/tt_metal/tt_metal/eth/test_eth_multi_txq_rxq.cpp:
```diff
@@ -0,0 +1,176 @@
+    uint32_t data_txq_id,
+    uint32_t ack_txq_id,
+    uint32_t num_messages) {
+    auto arch = tt::tt_metal::MetalContext::instance().hal().get_arch();
+    if (arch != tt::ARCH::BLACKHOLE) {
+        GTEST_SKIP();
+    }
```
This seems to be already covered in `P300DeviceFixture::SetUp`

daminakaTT commented on tt_metal/fabric/hw/inc/edm_fabric/compile_time_arg_tmp.hpp:
```diff
@@ -159,3 +159,17 @@ template <bool GET_THE_ARG, size_t ARG_IDX>
+template <bool GET_THE_ARG, typename T, size_t CT_ARGS_START_IDX, size_t NUM_ELEMS_TO_TAKE>
+constexpr auto conditional_get_next_n_args() -> typename std::enable_if<!GET_THE_ARG, std::array<T, NUM_ELEMS_TO_TAKE>>::type {
+    std::array<T, NUM_ELEMS_TO_TAKE> empty_arr{};
+    for (size_t i = 0; i < NUM_ELEMS_TO_TAKE; i++) {
```
I think `empty_arr = {}` automatically fill 0.
This loop is redundant

daminakaTT commented on tt_metal/fabric/hw/inc/edm_fabric/fabric_txq_setup.h:
```diff
@@ -0,0 +1,130 @@
+    // enable packet resend on each TXQ:
+    auto enable_packet_resend_txq = [](uint32_t addr) {
+        uint32_t val = 0;
+        val |= (1 << 0);  // packet resend
```
zero shift???

SeanNijjar replied:
just a stylistic choice to keep it consistent with the other lambda.

SeanNijjar replied:
Also I forgot to come back and revisit these and switch them out from literals. Moved to named constants now.

SeanNijjar committed:
WIP: add unit test to isolate bringup of two TXQ mode for BH in user kernels

update test to work around HW bug.

add RXQ REGS SIZE

WIP: Add changes to support new receiver channel -> sender channel credit scheme that is compatible with eth packet sends (as opoosed to register writes). This is required for 2 erisc  mode in BH because we can't send reg writes over two Ethernet txqs.

enable 2-txq mode with single risc on Blackhole

disable unsupported test case, cleanup kernel test code

cleanup

fix some bugs/accidental regressions

add missing licensing headers

rm dead code

relocate file to be consistent location with other headers

update include paths :facepalm:

add missing l1 cache invalidation

add missing data cache invalidation

address feedback/comments

sq
